src/infra/core/src/utils/simple_transfer_protocol.rs

```diff
@@ -247,130 +357,42 @@ impl SimpleTransferProtocol {
-                    .as_data_repo()
-                    .insert_stream(
-                        stream,
-                        InsertOpts {
-                            precomputed_hash: if !trust_source_hashes {
-                                None
-                            } else {
-                                Some(&data_slice.physical_hash)
-                            },
-                            expected_hash: Some(&data_slice.physical_hash),
-                            size_hint: Some(data_slice.size as usize),
-                            ..Default::default()
-                        },
-                    )
-                    .await
-                {
-                    Ok(_) => Ok(()),
-                    Err(InsertError::HashMismatch(e)) => Err(CorruptedSourceError {
-                        message: concat!(
-                            "Data file hash declared by the source didn't match ",
-                            "the computed - this may be an indication of hashing ",
-                            "algorithm mismatch or an attempted tampering",
-                        )
-                        .to_owned(),
-                        source: Some(e.into()),
-                    }
-                    .into()),
-                    Err(InsertError::Access(e)) => Err(SyncError::Access(e)),
-                    Err(InsertError::Internal(e)) => Err(SyncError::Internal(e)),
-                }?;
-
-                stats.src.data_slices_read += 1;
-                stats.dst.data_slices_written += 1;
-                stats.src.bytes_read += data_slice.size as usize;
-                stats.dst.bytes_written += data_slice.size as usize;
-                listener.on_status(SyncStage::TransferData, &stats);
+        let mut block_download_tasks = vec![];
+        blocks.iter().rev().for_each(|(_, b)| {
+            if let Some(block_stream) = b.as_data_stream_block() {
+                if let Some(data_slice) = block_stream.event.output_data {
+                    block_download_tasks.push(Box::pin(self.download_block_data(
+                        src,
+                        dst,
+                        data_slice,
+                        trust_source_hashes,
+                        listener.clone(),
+                        stats.clone(),
+                    ))
+                    // Each function return unique future
+                    // cast future to next type to allow storing them in vector
+                        as Pin<Box<dyn Future<Output = Result<(), SyncError>> + Send>>)
+                }
+                if let Some(checkpoint) = block_stream.event.output_checkpoint {
+                    block_download_tasks.push(Box::pin(self.download_block_checkpoints(
+                        src,
+                        dst,
+                        checkpoint,
+                        trust_source_hashes,
+                        listener.clone(),
+                        stats.clone(),
```

Comment:
I don't understand this cloning of `stats` - here and above.
How does the stats from N futures gets merged back?

In a similar place in Smart protocol, we are using atomic counters.